frontend/packages/core-auth/src/components/AuthWrapper.tsx

```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
     }
   }, [tokenResponse, isLoggedIn]);
 
+  // Schedule a token refresh before it expires (or use an existing schedule)
+  const refreshTokenTimeout = useRef<number | null>(null);
   useEffect(() => {
-    onTokenChange(token);
-  }, [token]);
+    if (
+      tokenResponse != null &&
+      tokenResponse.expiresIn != null &&
+      tokenResponse.expiresIn > 0
+    ) {
+      if (refreshTokenTimeout.current != null)
+        window.clearTimeout(refreshTokenTimeout.current);
+
+      const delay = tokenResponse.getExpiryMs() - Date.now();
+      console.log({ delay });
+
+      if (delay <= 0) {
+        refreshToken();
+      } else {
+        refreshTokenTimeout.current = window.setTimeout(
+          () => refreshToken(),
+          delay,
+        );
+      }
+    }
+  }, [JSON.stringify(tokenResponse), discovery]);
```

Comment:
Same here. I think `discovery` is ok, but you should check against the bits of `tokenResponse` you're interested in.